src/NServiceBus.Core.Tests/Recoverability/SecondLevelRetries/DelayedRetryActionTests.cs

```diff
@@ -0,0 +1,132 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using DelayedDelivery;
+    using Extensibility;
+    using NServiceBus.Transports;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class DelayedRetryActionTests
+    {
+        const string TimeoutManagerAddress = "timeout handling endpoint";
+        const string EndpointInputQueue = "endpoint input queue";
+
+        [Test]
+        public async Task Should_dispatch_original_message_body()
+        {
+            var dispatcher = new FakeDispatcher();
+            var delayedRetryAction = new DelayedRetryAction(EndpointInputQueue, dispatcher);
+            var originalBody = Encoding.UTF8.GetBytes("original message body");
+            var incomingMessage = new IncomingMessage("messageId", new Dictionary<string, string>(), new MemoryStream(originalBody));
+            incomingMessage.Body = Encoding.UTF8.GetBytes("updated message body");
+
+            await delayedRetryAction.Retry(incomingMessage, TimeSpan.Zero, new ContextBag());
+
+            var outgoingMessage = dispatcher.TransportOperations.UnicastTransportOperations.Single();
+            Assert.That(outgoingMessage.Message.Body, Is.EqualTo(originalBody));
+        }
+
+        [Test]
+        public async Task Should_float_context_to_dispatcher()
+        {
+            var context = new ContextBag();
+            var dispatcher = new FakeDispatcher();
+            var delayedRetryAction = new DelayedRetryAction(EndpointInputQueue, dispatcher);
+            var incomingMessage = new IncomingMessage("messageId", new Dictionary<string, string>(), Stream.Null);
+
+            await delayedRetryAction.Retry(incomingMessage, TimeSpan.Zero, context);
+
+            Assert.That(dispatcher.ContextBag, Is.SameAs(context));
+        }
+
+        [Test]
+        public async Task When_native_delayed_delivery_should_add_delivery_constraint()
+        {
+            var dispatcher = new FakeDispatcher();
+            var delayedRetryAction = new DelayedRetryAction(EndpointInputQueue, dispatcher);
+            var delay = TimeSpan.FromSeconds(42);
+            var incomingMessage = new IncomingMessage("messageId", new Dictionary<string, string>(), Stream.Null);
+
+            await delayedRetryAction.Retry(incomingMessage, delay, new ContextBag());
+
+            var outgoingMessage = dispatcher.TransportOperations.UnicastTransportOperations.Single();
+            Assert.That(outgoingMessage.Destination, Is.EqualTo(EndpointInputQueue));
+            var deliveryConstraint = outgoingMessage.DeliveryConstraints.OfType<DelayDeliveryWith>().SingleOrDefault();
+            Assert.That(deliveryConstraint, Is.Not.Null);
+            Assert.That(deliveryConstraint.Delay, Is.EqualTo(delay));
+        }
+
+        [Test]
+        public async Task When_no_native_delayed_delivery_should_route_message_to_timeout_manager()
```

Comment:
these are headers which are needed by the timeout manager. That logic was copied from the `TimeoutManagerRoutingStrategy` that stuff belongs together, I don't think there is value in splitting that up.